File: app-java/src/main/java/tech/thdev/java_udemy_sample/adapter/SampleAdapter.java
package tech.thdev.java_udemy_sample.adapter;

import android.content.Context;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

import tech.thdev.java_udemy_sample.adapter.holder.SampleViewHolder;
import tech.thdev.java_udemy_sample.data.SampleItem;

/**
 * Created by tae-hwan on 10/29/16.
 */

public class SampleAdapter extends RecyclerView.Adapter<SampleViewHolder> {

    private final Context context;
    private ArrayList<SampleItem> imageItems = new ArrayList<>();

    public SampleAdapter(Context context) {
        this.context = context;
    }

    @NotNull
    @Override
    public SampleViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        return new SampleViewHolder(context, parent);
    }

    @Override
    public void onBindViewHolder(SampleViewHolder holder, int position) {
        holder.onBindView(imageItems.get(position), position);
    }

    @Override
    public int getItemCount() {
        return imageItems.size();
    }

    public void addItem(SampleItem imageItem) {
        imageItems.add(imageItem);
    }
}

File: app-kotlin/src/main/java/tech/thdev/kotlin_udemy_sample/view/main/SampleAdapter.kt
package tech.thdev.kotlin_udemy_sample.view.main

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import tech.thdev.kotlin_udemy_sample.R

/**
 * Created by tae-hwan on 10/20/16.
 */

class SampleAdapter(private val context: Context) : RecyclerView.Adapter<SampleAdapter.SampleViewHolder>() {

    // new로 바로 생성
    private val itemList = ArrayList<Int>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SampleViewHolder {
        return SampleViewHolder(parent)
    }

    override fun onBindViewHolder(holder: SampleViewHolder, position: Int) {
        holder.bindView(itemList[position])
    }

    override fun getItemCount(): Int {
        return itemList.size
    }

    /**
     * Item을 추가한다
     */
    fun addItem(index: Int) {
        itemList.add(index)
    }

    inner class SampleViewHolder(parent: ViewGroup) :
            RecyclerView.ViewHolder(LayoutInflater.from(context).inflate(R.layout.item_sample, parent, false)) {

        private val textView by lazy {
            itemView.findViewById(R.id.text) as TextView
        }

//        init {
//            textView = itemView.findViewById(R.id.text) as TextView
//        }

        fun bindView(index: Int) {
            textView.text = "Index : $index"
        }
    }
}
